src/fides/api/service/connectors/dynamic_erasure_email_connector.py

```diff
@@ -0,0 +1,502 @@
+    collection_address: CollectionAddress
+    collection_data: Any
+    email_field: str
+    vendor_field: str
+    dsr_field_to_collection_field: Dict[str, str]
+
+
+class BatchedIdentitiesData(NamedTuple):
+    email_address: str
+    vendor_name: str
+    identities: List[str]
+
+
+class DynamicErasureEmailConnector(BaseErasureEmailConnector):
+    """Email Erasure Connector that performs a lookup for the email to use."""
+
+    config: DynamicErasureEmailSchema
+
+    def get_config(self, configuration: ConnectionConfig) -> DynamicErasureEmailSchema:
+        return DynamicErasureEmailSchema(**configuration.secrets or {})
+
+    def batch_email_send(self, privacy_requests: Query) -> None:
+        logger.debug(
+            "Starting batch_email_send for connector: {} ...", self.configuration.name
+        )
+
+        db: Session = Session.object_session(self.configuration)
+        proccessed_config = self.process_connector_config(db, privacy_requests)
+
+        skipped_privacy_requests: List[str] = []
+
+        # We'll batch identities for each (email address, vendor_name) pair
+        # the keys of the dictionary will be email_address:vendor_name
+        batched_identities: Dict[str, BatchedIdentitiesData] = {}
+
+        for privacy_request in privacy_requests:
+            try:
+                custom_field_data = (
+                    self.get_email_and_vendor_from_custom_request_fields(
+                        db,
+                        proccessed_config,
+                        privacy_request,
+                    )
+                )
+
+                # if we couldn't retrieve the email address / vendor, we skip the privacy request
+                # (error was already logged within the function, so we just continue to the
+                # next privacy request)
+                if not custom_field_data:
+                    continue
+
+                user_identities: Dict[str, Any] = (
+                    privacy_request.get_cached_identity_data()
+                )
+                filtered_user_identities: Dict[str, Any] = (
+                    filter_user_identities_for_connector(self.config, user_identities)
+                )
+
+                # If there are no user identities, we just skip the privacy request
+                if not filtered_user_identities:
+                    skipped_privacy_requests.append(privacy_request.id)
+                    self.add_skipped_log(db, privacy_request)
+                    continue
+
+                email, vendor_name = custom_field_data
+
+                key = f"{email}:{vendor_name}"
+                if key not in batched_identities:
+                    batched_identities[key] = BatchedIdentitiesData(
+                        email, vendor_name, []
+                    )
+
+                batched_identities[key].identities.extend(
+                    filtered_user_identities.values()
+                )
+
+            except Exception as exc:
```

Comment:
that's a good callout. The only caveat I see is that since this could run against any connector, we'd need to catch all possible connection errors (i.e from sqlalchemy, mongo, dynamo, etc etc). The good thing is right now we're only supporting this for SQL connectors (and Snowflake which we actually query with the SQLQueryConfig) so I can add the ones from that connector now. If we ever add support for other connectors we'd also need to make sure to catch their exceptions

Reply:
That's true.. there's also been some back and forth about the tradeoffs with this pattern elsewhere because we make it harder to roubleshoot - I expect us to continue to focus here -